core/IssueQueue.hpp

```diff
@@ -0,0 +1,112 @@
+  // instructions until an available ExecutePipe becomes ready Dispatch -> Issue
+  // Queue -> ExecutePipe
+public:
+  class IssueQueueParameterSet : public sparta::ParameterSet {
+  public:
+    IssueQueueParameterSet(sparta::TreeNode *n) : sparta::ParameterSet(n) {}
+    PARAMETER(uint32_t, scheduler_size, 8, "Scheduler queue size")
+    PARAMETER(bool, in_order_issue, true, "Force in order issue")
+  };
+
+  IssueQueue(sparta::TreeNode *node, const IssueQueueParameterSet *p);
+  static const char name[];
+  void setExePipe(std::string exe_pipe_name, olympia::ExecutePipe *exe_pipe);
+  void setExePipeMapping(InstArchInfo::TargetPipe tgt_pipe,
+                         olympia::ExecutePipe *exe_pipe);
+  std::unordered_map<std::string, olympia::ExecutePipe *> getExePipes();
+
+private:
+  // Scoreboards
+  using ScoreboardViews = std::array<std::unique_ptr<sparta::ScoreboardView>,
+                                     core_types::N_REGFILES>;
+  ScoreboardViews scoreboard_views_;
+
+  // Tick events
+  sparta::SingleCycleUniqueEvent<> ev_issue_ready_inst_{
+      &unit_event_set_, "issue_event",
+      CREATE_SPARTA_HANDLER(IssueQueue, sendReadyInsts_)};
+  void setupIssueQueue_();
+  void getInstsFromDispatch_(const InstPtr &);
+  void flushInst_(const FlushManager::FlushingCriteria &criteria);
+  void sendReadyInsts_();
+  void readyExeUnit_(const uint32_t &);
+  void appendIssueQueue_(const InstPtr &);
+  void popIssueQueue_(const InstPtr &);
+  void handleOperandIssueCheck_(const InstPtr &);
+  void onROBTerminate_(const bool &val);
+  void onStartingTeardown_() override;
+  void dumpDebugContent_(std::ostream &output) const override final;
+
+  sparta::DataInPort<InstQueue::value_type> in_execute_inst_{
+      &unit_port_set_, "in_execute_write", 1};
+  // in_exe_pipe_done tells us an execution pipe is done
+  sparta::DataInPort<uint32_t> in_exe_pipe_done_{&unit_port_set_,
+                                                 "in_execute_pipe"};
+  sparta::DataOutPort<uint32_t> out_scheduler_credits_{&unit_port_set_,
+                                                       "out_scheduler_credits"};
+  sparta::DataInPort<FlushManager::FlushingCriteria> in_reorder_flush_{
+      &unit_port_set_, "in_reorder_flush", sparta::SchedulingPhase::Flush, 1};
+  // mapping used to lookup based on execution unit name
+  // {"alu0": ExecutePipe0}
+  std::unordered_map<std::string, olympia::ExecutePipe *> exe_pipes_;
+  // mapping of target pipe -> vector of execute pipes
+  // i.e {"INT", <ExecutePipe0, ExecutePipe1>}
+  std::unordered_map<InstArchInfo::TargetPipe,
+                     std::vector<olympia::ExecutePipe *>>
+      pipe_exe_pipe_mapping_;
+  // storage of what pipes are in this issue queues
+  std::vector<olympia::ExecutePipe *> pipes_;
+  std::vector<std::string> exe_unit_str_;
+  const core_types::RegFile reg_file_;
+
+  // Ready queue
+  typedef std::list<InstPtr> ReadyQueue;
```

Comment:
Consider using a `PriorityQueue`: https://github.com/sparcians/map/blob/master/sparta/sparta/resources/PriorityQueue.hpp

Reply:
Resolved in [6b0e363](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/6b0e3633c1b7331170afda8920f596f9fd16b233)